### ISSUE TITLE
Sequence Tagging custom dataset

### ISSUE THREAD

**Issue by nth-attempt** [pytorch/text#860]
## ❓ Questions and Help

**Description**
I have a sequence tagging dataset with text and label spans. E.g.
`text = 'Apple is located\nin Palo Alto'`
`label_spans = [(0, 5, 'Company'), (20, 29, 'Location')]`

What is the best approach to create a Dataset and a DataLoader?

I also have a function to pre-process labels into their corresponding BIO labels. Where would I pre-process the labels? 

Any guidance will help. Thanks!

**Comment by nth-attempt** [comment]
Thank you. Is my understanding correct that 'raw' folder has IterableDatasets downloading from a URL and the main folder has Datasets from local paths?

**Comment by zhangguanheng66** [comment]
Yes. The `raw` dataset download, unzip files and generate `IterableDataset`. From the link above, you materialize the `IterableDataset` and generate the dataset.

**Comment by nth-attempt** [comment]
Nice. Is the idea to give more control to the user? How would this fit in with the existing `Field` class?  With the new pipeline, where would be batching, padding be handled? Also, if we want to add special tokens like '<sos>', '<eos>' to the vocab, I guess we incorporate them while building our `Vocab`?

**Comment by zhangguanheng66** [comment]
`Field` class will be retired and users should use `DataLoader` for batching and `collate_func` for padding. And yes, you will have to build vocab by yourself. Here is the issue to review the new abstraction https://github.com/pytorch/text/issues/664

**Comment by nth-attempt** [comment]
Awesome. Thanks!

**Comment by nth-attempt** [comment]
How do I go about implementing char and word level features in the above example? Would I use two Vocab

**Comment by nth-attempt** [comment]
How do we go about using char level and word level features in the new pipeline?

**Comment by akurniawan** [comment]
@nth-attempt I wrote [this example](https://gist.github.com/akurniawan/30719686669dced49e7ced720329a616) for machine translation for using char level features, hopefully can give you some idea for using both char and word features

**Comment by zhangguanheng66** [comment]
@akurniawan, do you think it's worthy to merge the example to the folder `text/examples`?

**Comment by akurniawan** [comment]
sure, will create a PR for that

**Comment by nth-attempt** [comment]
@akurniawan thanks for the example. So, I believe I could have `char_vocab`, `char_transforms`, `word_vocab`, `word_transforms` in the `Dataset` class and return both `char_tensor` and `word_tensor` on `get_item`. At that point, I am not using anything from the `torchtext` library other than `Vocab` class right?

**Comment by zhangguanheng66** [comment]
I believe you can also use SentencePiece to get the subword level information. Some utils are available [here](https://github.com/pytorch/text/blob/bcb9104680eb9dc978a6bbcc2b9ca46cf2bdbed9/torchtext/data/functional.py#L19-L107)

**Comment by akurniawan** [comment]
@nth-attempt not necessarily, you can also utilize some of the functionalities in torchtext library such as `vocab_func` for transforming the word level and as @zhangguanheng66 said `SentencePiece` tokenizer for the subword level. I needed to create separate function for the character level as I felt the current function can't handle nested dimension as what happened in the char level.
If you want, you can try to look at `text/torchtext/data/functional.py` and `text/torchtext/experimental/` for more active development in new torchtext API

**Comment by nth-attempt** [comment]
Got it. Will be watching that space. Thanks!

**Comment by zhangguanheng66** [comment]
Here is an example to use sentencepiece API for text classification dataset [link](https://github.com/pytorch/text/blob/master/examples/text_classification/spm_dataset.py). I'm thinking about a new example or an updated one for spm.
